lib/worker/batcher.ex

```diff
@@ -466,92 +436,25 @@ defmodule BorsNG.Worker.Batcher do
-                # If a user doesn't have a public email address in their GH profile
-                # then get the email from the first commit to the PR
-                user_email =
-                  if user.email != nil do
-                    user.email
-                  else
-                    Enum.at(commits, 0).author_email
-                  end
-
-                # The head sha is the final commit in the PR.
-                source_sha = pr.head_sha
-                Logger.info("Staging branch #{stmp}")
-                Logger.info("Commit sha #{source_sha}")
-
-                # Create a merge commit for each PR
-                # because each PR is merged on top of each other in stmp, we can verify against any merge conflicts
-                merge_commit =
-                  GitHub.merge_branch!(
-                    repo_conn,
-                    %{
-                      from: source_sha,
-                      to: stmp,
-                      commit_message:
-                        "[ci skip][skip ci][skip netlify] -bors-staging-tmp-#{source_sha}"
-                    }
-                  )
-
-                Logger.info("Merge Commit #{inspect(merge_commit)}")
-
-                Logger.info("Previous Head #{inspect(prev_head)}")
-
-                # Then compress the merge commit into tree into a single commit
-                # appent it to the previous commit
-                # Because the merges are iterative the contain *only* the changes from the PR vs the previous PR(or head)
-
-                commit_message =
-                  Batcher.Message.generate_squash_commit_message(
-                    pr,
-                    commits,
-                    user_email,
-                    toml.cut_body_after
-                  )
-
-                cpt =
-                  GitHub.create_commit!(
-                    repo_conn,
-                    %{
-                      tree: merge_commit.tree,
-                      parents: [prev_head],
-                      commit_message: commit_message,
-                      committer: %{name: user.name || user.login, email: user_email}
-                    }
-                  )
-
-                Logger.info("Commit Sha #{inspect(cpt)}")
-                cpt
-              end)
+          cond do
+            squash? and local? ->
+              %{commit: commit} = Merge.Local.squash_merge_batch!(batch, patch_links, toml)
+              [commit]
 
-            GitHub.delete_branch!(repo_conn, stmp)
-            [new_head]
-          else
-            parents = [base.commit | Enum.map(patch_links, & &1.patch.commit)]
-            parents
+            squash? and not local? ->
+              commit = Merge.API.squash_merge_batch!(batch, patch_links, base, toml)
+              [commit]
+
+            not squash? ->
+              [base.commit | Enum.map(patch_links, & &1.patch.commit)]
           end
 
         head =
-          if toml.use_squash_merge do
+          if squash? do
```

Comment:
below here is where Bors is actually pushing staging to the merge result. again, note that Bors manually constructs the merge commit with `synthesize_commit!()` so that there's only a single merge commit per batch (unless squash merges are enabled)